common/src/main/java/com/viaversion/viabackwards/protocol/protocol1_16_4to1_17/packets/BlockItemPackets1_17.java

```diff
@@ -82,24 +83,85 @@ public void registerMap() {
+                        state.setLastCursorItem(clicked.copy());
+                    } else if (mode == 0 && button == 1 && clicked != null) {
+                        // Right click PICKUP
+                        // Carried item will (usually) be half of the clicked stack (rounding up)
+                        // if the clicked slot is empty, otherwise it will (usually) be the whole clicked stack
+                        state.setLastCursorItem(clicked.copy());
+                        if (state.getLastCursorItem() == null) {
+                            state.getLastCursorItem().setAmount((clicked.amount() + 1) / 2);
+                        }
+                    } else if (mode == 5 && slot == -999 && (button == 0 || button == 4)) {
+                        // Start drag (left or right click)
+                        // Preserve guessed carried item and forward to server
+                        // This mostly fixes the click and drag ghost item issue
+
+                        // no-op
+                    } else {
+                        // Carried item unknown (TODO)
+                        state.setLastCursorItem(null);
+                    }
+
+                    Item carried = state.getLastCursorItem();
+                    if (carried == null) {
+                        // Expected is the carried item after clicking, old clients send the clicked one (*mostly* being the same)
+                        wrapper.write(Type.FLAT_VAR_INT_ITEM, clicked);
+                    } else {
+                        wrapper.write(Type.FLAT_VAR_INT_ITEM, carried);
+                    }
                 });
             }
         });
+
+        protocol.registerClientbound(ClientboundPackets1_17.SET_SLOT, new PacketRemapper() {
+            @Override
+            public void registerMap() {
+                handler(wrapper -> {
+                    short windowId = wrapper.passthrough(Type.UNSIGNED_BYTE);
+                    short slot = wrapper.passthrough(Type.SHORT);
+
+                    Item carried = wrapper.read(Type.FLAT_VAR_INT_ITEM);
+                    if (carried != null && windowId == -1 && slot == -1) {
+                        // This is related to the hack to fix click and drag ghost items above.
+                        // After a completed drag, we have no idea how many items remain on the cursor,
+                        // and vanilla logic replication would be required to calculate the value.
+                        // When the click drag complete packet is sent, we will send an incorrect
+                        // carried item, and the server will helpfully send this packet allowing us
+                        // to update the internal state. This is necessary for fixing multiple sequential
+                        // click drag actions without intermittent pickup actions.
+                        wrapper.user().get(PlayerLastCursorItem.class).setLastCursorItem(handleItemToServer(carried.copy()));
```

Comment:
You probably should not call handleItemToServer() here, as that method remaps items in client -> server direction, which set slot is not. You should instead be storing the item which is returned from handleItemToClient() a couple of lines below. This ensures your tracker has the 1.16.4 item tracked.